src/Symfony/Component/Config/Definition/EnumNode.php

```diff
@@ -78,13 +104,48 @@ protected function finalizeValue(mixed $value): mixed
     {
         $value = parent::finalizeValue($value);
 
-        if (!\in_array($value, $this->values, true)) {
-            $ex = new InvalidConfigurationException(\sprintf('The value %s is not allowed for path "%s". Permissible values: %s', json_encode($value), $this->getPath(), $this->getPermissibleValues(', ')));
-            $ex->setPath($this->getPath());
+        if ($this->enumFqcn) {
+            if (is_a($this->enumFqcn, \BackedEnum::class, true)) {
+                if (\is_string($value) || \is_int($value)) {
+                    try {
+                        $case = $this->enumFqcn::tryFrom($value);
+                    } catch (\TypeError) {
+                        throw new InvalidConfigurationException(\sprintf('The value could not be casted to a case of the "%s" enum. Is the value the same type as the backing type of the enum?', $this->enumFqcn));
+                    }
+
+                    if (null !== $case) {
+                        return $case;
+                    }
+                } elseif ($value instanceof \UnitEnum && !$value instanceof $this->enumFqcn) {
+                    throw new InvalidConfigurationException(\sprintf('The value should be part of the "%s" enum, got a value from the "%s" enum.', $this->enumFqcn, get_debug_type($value)));
+                }
```

Comment:
Why are we only performing this check if the configured enum is a backed one?

Reply:
UnitEnum checks is straightforward because we don't need to cast values or check if they are a case value and so on. Structuring the code like this allows to have a quicker path in case of a UnitEnum with only one `instanceof` to check.